File: compiler/src/tir/qualified_name.rs
use config::SOURCE_EXT;
use std::path::MAIN_SEPARATOR;

#[derive(Debug)]
pub struct QualifiedName {
    pub parts: Vec<String>,
}

impl QualifiedName {
    pub fn new(parts: Vec<String>) -> Self {
        assert!(parts.len() > 0);

        QualifiedName { parts: parts }
    }

    pub fn module_name(&self) -> &String {
        self.parts.last().unwrap()
    }

    pub fn source_path_with_extension(&self) -> String {
        self.parts.join(&MAIN_SEPARATOR.to_string()) + SOURCE_EXT
    }
}

File: compiler/lib/inkoc/tir/qualified_name.rb
# frozen_string_literal: true

module Inkoc
  module TIR
    class QualifiedName
      attr_reader :parts

      def initialize(parts)
        if parts.empty?
          raise ArgumentError, 'Qualified names must contain at least 1 segment'
        end

        @parts = parts
      end

      def module_name
        @parts.last
      end

      def source_path_with_extension
        @parts.join(File::SEPARATOR) + Config::SOURCE_EXT
      end

      def to_s
        @parts.join(Config::MODULE_SEPARATOR)
      end

      def inspect
        "QualifiedName(#{self})"
      end
    end
  end
end
